### ISSUE TITLE
Check AppDomain before loading Extension Assembly

### ISSUE THREAD

**Issue by AchimStuy** [NLog/NLog#3678]
In [`AssemblyHelpers.LoadFromName`](https://github.com/NLog/NLog/blob/dev/src/NLog/Internal/AssemblyHelpers.cs#L105) please check in AppDomain, whether the file is already loaded. 

We are developing Addins for AutoCAD and have a custom Target in an extra Library, which would be in every of our Addins. Specifying `assemblyName` in `Extensions`-Section of nlog.config leads to a `FileNotFoundException`, if nlog is already loaded by another Addin (NLog is looking into directory of nlog.dll (the other addin's directory) and in AutoCAD-Installation-Directory, but not in my addin's directory).
We cannot foresee, where the Autodesk or the User installs the Addin, so specifying Extension via assemblyFile is not an option. But I am able to load my Assembly into AppDomain.

**NLog version**: 4.6.8
**Platform**: .Net 4.5

**Current NLog config:**
```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      internalLogLevel="Debug"
      internalLogFile="C:\Temp\nlog-internal.log"
      >

  <extensions>
    <add assembly="AutoCAD.Extensions.NLog" />
  </extensions>
  
  <targets>
    <target name="autocadConsole" xsi:type="AutoCADConsole" />
  </targets>

  <rules>
    <logger name="MyAddin" minlevel="Info" writeTo="autocadConsole" />
  </rules>
</nlog>
```

**Current Result**:
Error loading extensions. Exception: System.IO.FileNotFoundException: Could not load file or assembly 'AutoCAD.Extensions.NLog' or one of its dependencies.

**Expected Result**:
NLog is using the Assembly already loaded into AppDomain.

**Solution**:
Change [`AssemblyHelpers.LoadFromName`](https://github.com/NLog/NLog/blob/dev/src/NLog/Internal/AssemblyHelpers.cs#L105) as follows:
```C#
var name = new AssemblyName(assemblyName);
var loadedAssembly = AppDomain.CurrentDomain.GetAssemblies().FirstOrDefault(domainAssembly =>
{
    var domainAssemblyName = domainAssembly.GetName();
    if (name.Name != domainAssemblyName.Name)
        return false;
    if (name.Version != domainAssemblyName.Version)
        return false;
    if (name.CultureInfo != null && name.CultureInfo != domainAssemblyName.CultureInfo)
        return false;
    if (!name.GetPublicKeyToken().SequenceEqual(domainAssemblyName.GetPublicKeyToken()))
        return false;
    return true;
});
return loadedAssembly ?? Assembly.Load(assemblyName);
```

**Full Stacktrace**:
2019-11-07 10:42:21.5066 Error Error loading extensions. Exception: System.IO.FileNotFoundException: Could not load file or assembly 'AutoCAD.Extensions.NLog' or one of its dependencies. Das System kann die angegebene Datei nicht finden.
File name: 'AutoCAD.Extensions.NLog'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection)
   at System.Reflection.Assembly.Load(String assemblyString)
   at NLog.Internal.AssemblyHelpers.LoadFromName(String assemblyName)
   at NLog.Config.XmlLoggingConfiguration.ParseExtensionWithAssembly(String assemblyName, String prefix)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].

**Comment by snakefoot** [comment]
You are welcome to create a pull-requests the starts with doing:

`System.Reflection.Assembly.Load(String assemblyString)`

And if that fails with `FileNotFoundException`, then you can fallback to:

```
var name = new AssemblyName(assemblyName);
var loadedAssembly = AppDomain.CurrentDomain.GetAssemblies().FirstOrDefault(domainAssembly =>
{
    var domainAssemblyName = domainAssembly.GetName();
    if (name.Name != domainAssemblyName.Name)
        return false;
    if (name.Version != domainAssemblyName.Version)
        return false;
    if (name.CultureInfo != null && name.CultureInfo != domainAssemblyName.CultureInfo)
        return false;
    if (!name.GetPublicKeyToken().SequenceEqual(domainAssemblyName.GetPublicKeyToken()))
        return false;
    return true;
});
```

**Comment by AchimStuy** [comment]
Okay, I will create PR next week or the week after.
But why that way round? Why not first check AppDomain and then Load Assembly?
I also have no idea how to unit test this, but let's see.

Sleeping over it, I instead had the idea of checking not only nlog.dll-Directory and Application-Directory during load of NLog-extensions but also nlog.config-Directory. But I don't know, where this is implemented. Or is this .net doing for us? What do you think about this?

**Comment by snakefoot** [comment]
> But why that way round? 

Because 1 application out of 100.000 applications needs additional work to operate, then it doesn't mean the 100.000 should have the penalty of saving the 1 application.

**Comment by AchimStuy** [comment]
Where did you get the specs defined in [Device Specific Metadata](https://github.com/imbushuo/mac-precision-touchpad/tree/master/src/AmtPtpDeviceUsbUm/DeviceFamily)? I have added my device ID and it's working ok, but I would like to fix the size information.

**Comment by snakefoot** [comment]
@AchimStuy What happens if you do this, instead of scanning manually:

`Assembly.Load(new AssemblyName(assemblyName));`

Or this:

`Assembly.Load(AssemblyName.GetAssemblyName(assemblyName));`

**Comment by 304NotModified** [comment]
@AchimStuy 

please note: master is targeting NLog 4.7 and dev 5.0

> Sleeping over it, I instead had the idea of checking not only nlog.dll-Directory and Application-Directory during load of NLog-extensions but also nlog.config-Directory. But I don't know, where this is implemented.

I think you're looking for this:

https://github.com/NLog/NLog/blob/master/src/NLog/Config/ConfigurationItemFactory.cs#L431-L464

**Comment by AchimStuy** [comment]
> @AchimStuy What happens if you do this, instead of scanning manually:
> 
> `Assembly.Load(new AssemblyName(assemblyName));`
> 
> Or this:
> 
> `Assembly.Load(AssemblyName.GetAssemblyName(assemblyName));`

Both not working.

> > Sleeping over it, I instead had the idea of checking not only nlog.dll-Directory and Application-Directory during load of NLog-extensions but also nlog.config-Directory. But I don't know, where this is implemented.
> 
> I think you're looking for this:
> 
> https://github.com/NLog/NLog/blob/master/src/NLog/Config/ConfigurationItemFactory.cs#L431-L464

This is also not working, because the code is executed during static creation of default `ConfigurationItemFactory`-Instance. But I would need the nlog.config-path.

So I will implement as [proposed](https://github.com/NLog/NLog/issues/3678#issuecomment-551176312) by @snakefoot.

> please note: master is targeting NLog 4.7 and dev 5.0

I would implement on master. Okay, @304NotModified?

**Comment by 304NotModified** [comment]
> This is also not working, because the code is executed during static creation of default `ConfigurationItemFactory`-Instance. But I would need the nlog.config-path.

true, I was thinking your were looking for the current code

> I would implement on master. Okay, @304NotModified?

👍

**Comment by snakefoot** [comment]
Resolved with #3679 ?

**Comment by 304NotModified** [comment]
Yes, unfortunately GitHub won't close it automatically as it isn't merged to the default branch (yet)

Thanks for the reminder
